In [210]:
A = [1,2,3,4,5,6]
B = [13,21,34]

In [211]:
A.extend(B)
A

[1, 2, 3, 4, 5, 6, 13, 21, 34]

IDENTITY MATRIX

a square matrix in which all the elements of the principal diagonal are ones and all other elements are zeros. The effect of multiplying a given matrix by an identity matrix is to leave the given matrix unchanged

In [212]:
import numpy as np
np.identity(4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

In [213]:
np.array([(1,0,0),(0,1,0),(0,0,1)])

array([[1, 0, 0],
       [0, 1, 0],
       [0, 0, 1]])

In [214]:
np.eye(3)

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [215]:
import pandas as pd
data = pd.read_csv('fuel_ferc1.csv')

In [216]:
data.head()

record_id  utility_id_ferc1  report_year  \
0   f1_fuel_1994_12_1_0_7                 1         1994   
1  f1_fuel_1994_12_1_0_10                 1         1994   
2   f1_fuel_1994_12_2_0_1                 2         1994   
3   f1_fuel_1994_12_2_0_7                 2         1994   
4  f1_fuel_1994_12_2_0_10                 2         1994   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
0              rockport                coal       ton        5377489.0   
1  rockport total plant                coal       ton       10486945.0   
2                gorgas                coal       ton        2978683.0   
3                 barry                coal       ton        3739484.0   
4             chickasaw                 gas       mcf          40533.0   

   fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
0               16.590                      18.59   
1               16.592                      18.58   
2               24.130                      39.72   
3               23.950                      47.21   
4                1.000                       2.77   

   fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
0                         18.53                1.121  
1                         18.53                1.120  
2                         38.12                1.650  
3                         45.99                1.970  
4                          2.77                2.570

In [217]:
# to obtain unique values in a dataframe
data['fuel_type_code_pudl'].unique()

array(['coal', 'gas', 'nuclear', 'oil', 'waste', 'other'], dtype=object)

In [233]:
data.loc[data.fuel_type_code_pudl=='nuclear', 'fuel_cost_per_unit_burned'].describe()

count      818.000000
mean      4955.157002
std      14291.929749
min          0.000000
25%          1.601500
50%         32.520000
75%         50.855250
max      69785.215000
Name: fuel_cost_per_unit_burned, dtype: float64

In [219]:
for i in data['fuel_type_code_pudl'].unique():
    print('the mean for {} is : {}'.format(i, data.loc[data.fuel_type_code_pudl==i, 'fuel_cost_per_unit_burned'].describe()[1]))

the mean for coal is : 67.42182999882996
the mean for gas is : 13.659396569737135
the mean for nuclear is : 4955.157002444988
the mean for oil is : 168.87708568948383
the mean for waste is : 19.518122448979593
the mean for other is : 18.253856287425158


In [220]:
data['fuel_mmbtu_per_unit'].describe()

count    29523.000000
mean         8.492111
std         10.600220
min          0.000001
25%          1.024000
50%          5.762694
75%         17.006000
max        341.260000
Name: fuel_mmbtu_per_unit, dtype: float64

The skewness is a parameter to measure the symmetry of a data set. 
While the kurtosis is a parameter to measure how heavy its tails are compared to a normal distribution.

In [221]:
from scipy.stats import kurtosis, skew

x = data['fuel_qty_burned']

print('skewness: {}, kurtosis: {}'.format(skew(x), kurtosis(x)))

skewness: 15.850690077422112, kurtosis: 651.2589365474387


In [222]:
total_na = data.isna().sum()
total_na

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [223]:
(total_na/len(data))*100

record_id                       0.000000
utility_id_ferc1                0.000000
report_year                     0.000000
plant_name_ferc1                0.000000
fuel_type_code_pudl             0.000000
fuel_unit                       0.609694
fuel_qty_burned                 0.000000
fuel_mmbtu_per_unit             0.000000
fuel_cost_per_unit_burned       0.000000
fuel_cost_per_unit_delivered    0.000000
fuel_cost_per_mmbtu             0.000000
dtype: float64

Feature: fuel_unit, Percent: 0.610%

Categorical and mode imputation

The statistical relationship between two variables is referred to as their correlation

correlation can be positive (both change in same direction), negative (both change in opposite direction, or neutral (no relationship in the change of the variables)

In [224]:
# DataFrame.corr() is used to find the pairwise correlation of all columns in the dataframe.

data.corr()

utility_id_ferc1  report_year  fuel_qty_burned  \
utility_id_ferc1                      1.000000     0.093323        -0.057447   
report_year                           0.093323     1.000000         0.012952   
fuel_qty_burned                      -0.057447     0.012952         1.000000   
fuel_mmbtu_per_unit                  -0.066946    -0.110853        -0.080946   
fuel_cost_per_unit_burned            -0.037863     0.013599        -0.018535   
fuel_cost_per_unit_delivered         -0.016414    -0.014043        -0.003551   
fuel_cost_per_mmbtu                   0.006122     0.010261        -0.001896   

                              fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
utility_id_ferc1                        -0.066946                  -0.037863   
report_year                             -0.110853                   0.013599   
fuel_qty_burned                         -0.080946                  -0.018535   
fuel_mmbtu_per_unit                      1.000000                  -0.010034   
fuel_cost_per_unit_burned               -0.010034                   1.000000   
fuel_cost_per_unit_delivered            -0.009039                   0.011007   
fuel_cost_per_mmbtu                     -0.005884                  -0.000437   

                              fuel_cost_per_unit_delivered  \
utility_id_ferc1                                 -0.016414   
report_year                                      -0.014043   
fuel_qty_burned                                  -0.003551   
fuel_mmbtu_per_unit                              -0.009039   
fuel_cost_per_unit_burned                         0.011007   
fuel_cost_per_unit_delivered                      1.000000   
fuel_cost_per_mmbtu                              -0.000109   

                              fuel_cost_per_mmbtu  
utility_id_ferc1                         0.006122  
report_year                              0.010261  
fuel_qty_burned                         -0.001896  
fuel_mmbtu_per_unit                     -0.005884  
fuel_cost_per_unit_burned               -0.000437  
fuel_cost_per_unit_delivered            -0.000109  
fuel_cost_per_mmbtu                      1.000000

fuel_qty_burned & fuel_mmbtu_per_unit


In [234]:
data.groupby(data['report_year'])['fuel_cost_per_unit_delivered'].transform('mean')

0         63.636060
1         63.636060
2         63.636060
3         63.636060
4         63.636060
            ...    
29518    499.269966
29519    499.269966
29520    499.269966
29521    499.269966
29522    499.269966
Name: fuel_cost_per_unit_delivered, Length: 29523, dtype: float64

In [226]:
w = data[data.fuel_type_code_pudl=='coal']

year_1994 = []
[year_1994.append(i) for i in w['report_year']==1994]
sum_1 = sum(year_1994)
 
year_1998 = []
[year_1998.append(i) for i in w['report_year']==1998]
sum_2 = sum(year_1998)

In [227]:
(sum_1 - sum_2)/(sum_1)*100

9.263157894736842